include/ratelimit_manager.h

```diff
@@ -0,0 +1,174 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include <tuple>
+#include <unordered_map>
+#include <mutex>
+#include <shared_mutex>
+#include "lru/lru.hpp"
+#include "option.h"
+
+
+
+
+// Action enum for rate limit rules
+enum class RateLimitAction {
+    ALLOW,
+    BLOCK,
+    THROTTLE
+};
+
+enum class RateLimitedResourceType {
+    IP,
+    API_KEY
+};
+
+struct rate_limit_throttle_t {
+    int64_t minute_rate_limit;
+    int64_t hour_rate_limit;
+
+};
+
+struct rate_limit_rule_t {
+    uint64_t id;
+    RateLimitAction action;
+    RateLimitedResourceType resource_type;
+    std::vector<std::string> values;
```

Comment:
Calling this as `entity_ids` is probably more clearer than `values` and goes well with `entity_type` above.